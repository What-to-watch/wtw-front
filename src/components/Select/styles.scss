.select {
    position: relative;

    &__button{
        display: flex;
        align-items: center;
        & span {
            color: white;
            font-weight: bold;
            font-size: 16px;
            line-height: 19px;
            margin-right: 3px;
        }
        & img {
            transition: transform ease-out 150ms;
            transform: rotate(0deg)
        }
    }

    & .list{
        position: absolute;
        bottom: 100%;
        display: hidden;
        z-index: -100;
        transition: opacity ease-out 150ms;
        opacity: 0;
    }

    &.open {
        & .list{
            display: block;
            opacity: 1;
            z-index: 99;
        }

        & .select__button{
            & img {
                transform: rotate(180deg);
            }
        }
    }
}