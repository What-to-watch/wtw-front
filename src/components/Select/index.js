import React, { useState } from 'react';
import PropTypes from 'prop-types';
import List from '../List';
import getClassName from '../../utils/getClassName';

import './styles.scss';

const Select = (props) => {
    const {initValue, options, open, onChange = x => x, onClick = x => x } = props;

    const selectOption = (value) => {
        return options.filter(option => option.value === value)[0]
    }

    const [ value, setValue ] = useState(initValue);

    const [ selected, setSelected ] = useState(
        selectOption(value)
    );

    const handleChange = (value, props) => {
        setValue(value);
        setSelected(selectOption(value));
        onChange(value, props);
    }

    const handleClick = (e) => {
        onClick(props);
    }

    const classes = getClassName({
        select: true,
        open: open,
    })

    return (
        <div className={classes}>
            <button onClick={handleClick} className="select__button">
                <span>{selected.label}</span>
                <img src="icons/down-arrow.svg" alt="down"/>
            </button>
            <List
                content={options}
                onItemClick={handleChange}
            />
        </div>
    )
}

Select.protoTypes = {
    initValue: PropTypes.string.isRequired,
    onChange: PropTypes.func,
    options: PropTypes.arrayOf({
        value: PropTypes.string,
        label: PropTypes.string,
    }).isRequired,
    open: PropTypes.bool,
    onClick: PropTypes.func
}

export default Select