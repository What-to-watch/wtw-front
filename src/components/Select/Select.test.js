import React from 'react';
import Select from './index';
import { render, fireEvent, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

describe ('Select render correctly', ()=>{
    const resultOptions = [
        { label: '10', value: 10 },
        { label: '100', value: 100 },
        { label: '1000', value: 1000 }
    ]
    let props = {
        options: resultOptions,
        initValue: 100,
        onChange: x => x,
        open: false,
    }
    const component = renderer.create(
        (<Select {...props}/>),
    );

    let select = component.toJSON();
    const labelNode = component.root.findByProps({className: "select__button"}).children[0];

    expect(labelNode.children[0]).toBe(props.initValue.toString());

    expect(select).toMatchSnapshot();

    it('Select open', ()=>{
        props = { ...props, open: true }
        const openComponent = renderer.create(
            (<Select {...props}/>)
        );

        const openSelect = openComponent.root.findByProps({className: "select open"}).children;

        expect(openSelect).toHaveLength(2);
    })

    it('Select onClick is executed with button', ()=>{
        const spy = (() => {
            let _spy = (...args) => {
                _spy.calls.push({ args })
                _spy.called = true;
                _spy.callCount = _spy.callCount+1;
            }
            _spy.calls = [];
            _spy.called = false;
            _spy.callCount = 0;
            Object.defineProperty(_spy,"firstCall",{
                get: () => _spy.calls[0]
            })
            return _spy
        })()

        props = { ...props, onClick: spy }

        render(<Select {...props} />);

        fireEvent.click(screen.getByRole('button'));

        expect(spy.called).toBeTruthy();
        expect(spy.callCount).toBe(1);
        expect(spy.firstCall.args[0]).toStrictEqual(props);
    });
})