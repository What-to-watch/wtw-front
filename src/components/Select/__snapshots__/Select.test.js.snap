// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  className="select"
>
  <button
    className="select__button"
    onClick={[Function]}
  >
    <span>
      100
    </span>
    <img
      alt="down"
      src="icons/down-arrow.svg"
    />
  </button>
  <ul
    className="list"
  >
    <li
      className="list__item"
      onClick={[Function]}
    >
      10
    </li>
    <li
      className="list__item"
      onClick={[Function]}
    >
      100
    </li>
    <li
      className="list__item"
      onClick={[Function]}
    >
      1000
    </li>
  </ul>
</div>
`;
